external-crates/move/crates/move-compiler/src/typing/translate.rs

```diff
@@ -2937,12 +2938,88 @@ fn module_call_impl(
     (call, return_)
 }
 
+/// If the constant that we are referencing has an `error` attribute, we need to change the type of
+/// the constant to a u64 since this will be compiled into a u64 error code.
+fn annotated_error_const(context: &mut Context, e: &mut T::Exp) {
+    let sp!(
+        const_loc,
+        T::UnannotatedExp_::Constant(module_ident, constant_name)
+    ) = &mut e.exp
+    else {
+        return;
+    };
+
+    let ConstantInfo {
+        attributes,
+        defined_loc,
+        signature,
+    } = context.constant_info(module_ident, constant_name);
+    let has_error_annotation = attributes.contains_key_(&known_attributes::ErrorAttribute.into());
+    let is_u64_type = matches!(
+        signature.value.builtin_name(),
+        Some(sp!(_, N::BuiltinTypeName_::U64))
+    );
+    let ty_loc = e.ty.loc;
+    let const_def_loc = *defined_loc;
+
+    if has_error_annotation {
+        context.env.check_feature(
+            context.current_package(),
+            FeatureGate::CleverAssertions,
+            *const_loc,
+        );
+    }
+
+    if context
+        .env
+        .supports_feature(context.current_package(), FeatureGate::CleverAssertions)
+    {
+        e.ty = N::Type_::builtin(ty_loc, sp(ty_loc, N::BuiltinTypeName_::U64), vec![]);
+    }
+
+    // Add help messages
+    if !has_error_annotation
+        && !is_u64_type
+            // If they're trying to use a non-u64 constant as an error code and in legacy,
+            // nudge them that this is available in Move 2024.
+            && context.env.check_feature(
+                context.current_package(),
+                FeatureGate::CleverAssertions,
+                *const_loc,
+                )
```

Comment:
I don't quite get the point of this. I think just erroring on the constant itself should be enough?

Reply:
This is for the case where in legacy they have code like this

```move
const B: vector<u8> = b"...";

fun f() {
   assert!(..., B);
}
```

This is then meant to give a "Clever assertions are available in Move 2024..." message.